File: fern/snippets/tool-calling/versioning/typescript/breaking-changes.ts
import { Composio } from "@composio/core";

const composio = new Composio({ apiKey: "YOUR_API_KEY" });

// ❌ This will fail without version specification (after October 22, 2025)
try {
    const result = await composio.tools.execute({
        toolSlug: "GITHUB_CREATE_ISSUE",
        userId: "user-123",
        arguments: {
            repo: "my-repo",
            title: "Bug report",
            body: "Issue description"
        }
    });
} catch (error) {
    console.error("Error:", error);
    console.log("Manual execution requires toolkit version");
}

// ✅ Solution 1: Specify version in execution
const result1 = await composio.tools.execute({
    toolSlug: "GITHUB_CREATE_ISSUE",
    userId: "user-123",
    arguments: {
        repo: "my-repo",
        title: "Bug report",
        body: "Issue description"
    },
    toolkitVersion: "20250116_00"
});

// ✅ Solution 2: Configure version at SDK level
const composioWithVersion = new Composio({
    apiKey: "YOUR_API_KEY",
    toolkitVersions: { github: "20250116_00" }
});

const result2 = await composioWithVersion.tools.execute({
    toolSlug: "GITHUB_CREATE_ISSUE",
    userId: "user-123",
    arguments: {
        repo: "my-repo",
        title: "Bug report",
        body: "Issue description"
    }
});

// ⚠️ Emergency escape hatch (not recommended for production)
const result3 = await composio.tools.execute({
    toolSlug: "GITHUB_CREATE_ISSUE",
    userId: "user-123",
    arguments: {
        repo: "my-repo",
        title: "Bug report",
        body: "Issue description"
    },
    dangerouslySkipVersionCheck: true
});

File: fern/snippets/tool-calling/versioning/python/breaking-changes.py
from composio import Composio

composio = Composio(api_key="YOUR_API_KEY")

# ❌ This will fail without version specification (after October 22, 2025)
try:
    result = composio.tools.execute(
        "GITHUB_CREATE_ISSUE",
        {
            "user_id": "user-123",
            "arguments": {
                "repo": "my-repo",
                "title": "Bug report",
                "body": "Issue description"
            }
        }
    )
except Exception as e:
    print(f"Error: {e}")
    print("Manual execution requires toolkit version")

# ✅ Solution 1: Specify version in execution
result = composio.tools.execute(
    "GITHUB_CREATE_ISSUE",
    {
        "user_id": "user-123",
        "arguments": {
            "repo": "my-repo",
            "title": "Bug report",
            "body": "Issue description"
        },
        "toolkit_version": "20250116_00"
    }
)

# ✅ Solution 2: Configure version at SDK level
composio_with_version = Composio(
    api_key="YOUR_API_KEY",
    toolkit_versions={"github": "20250116_00"}
)

result = composio_with_version.tools.execute(
    "GITHUB_CREATE_ISSUE",
    {
        "user_id": "user-123",
        "arguments": {
            "repo": "my-repo",
            "title": "Bug report",
            "body": "Issue description"
        }
    }
)

# ⚠️ Emergency escape hatch (not recommended for production)
result = composio.tools.execute(
    "GITHUB_CREATE_ISSUE",
    {
        "user_id": "user-123",
        "arguments": {
            "repo": "my-repo",
            "title": "Bug report",
            "body": "Issue description"
        },
        "dangerously_skip_version_check": True
    }
)